### PR TITLE
[WIP] Effects Of #26034 StringComparer Usage Changes.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
Issue #26034. **Test Failures Are To Be Expected**

Change some usage of `StringComparer.OrdinalIgnoreCase` and `CaseInse…nsitiveComparison.Comparer`
To `CaseInsensitiveComparison`

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs [hunk 1]
```diff
@@ -668,7 +668,7 @@ public override int GetHashCode()
         {
             return Hash.Combine(base.GetHashCodeHelper(),
                    Hash.Combine(this.AllowUnsafe,
-                   Hash.Combine(Hash.CombineValues(this.Usings, StringComparer.Ordinal),
+                   Hash.Combine(Hash.CombineValues(this.Usings, StringComparers.IdentifierComparer),
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpCompilationOptions.cs]: >IdentifierComparer [](start = 80, length = 18)

This is incorrect. In C#, identifier comparisons are not case insensitive.

File: src/Compilers/Core/Portable/StringComparers.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    public static class StringComparers
+    {
+        public static StringComparer FileSystemNameComparer =  StringComparer.OrdinalIgnoreCase;
+       public static StringComparer IdentifierComparer = CaseInsensitiveComparison.Comparer; 
+    }
```
Comment (gafter) [inline src/Compilers/Core/Portable/StringComparers.cs]: This is incorrect. In C#, identifier comparisons are not case insensitive.

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs [hunk 3]
```diff
@@ -160,7 +160,7 @@ protected override ParseOptions CommonWithFeatures(IEnumerable<KeyValuePair<stri
         public new CSharpParseOptions WithFeatures(IEnumerable<KeyValuePair<string, string>> features)
         {
             ImmutableDictionary<string, string> dictionary =
-                features?.ToImmutableDictionary(StringComparer.OrdinalIgnoreCase)
+                features?.ToImmutableDictionary(StringComparers.IdentifierComparer)
                 ?? ImmutableDictionary<string, string>.Empty;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: The is incorrect. Features are not intended to be compares case insensitively.

File: src/Compilers/CSharp/Portable/Binder/ImportChain.cs [hunk 4]
```diff
@@ -92,7 +92,7 @@ private ImmutableArray<Cci.UsedNamespaceOrType> TranslateImports(Emit.PEModuleBu
             {
                 var aliases = ArrayBuilder<string>.GetInstance(aliasSymbols.Count);
                 aliases.AddRange(aliasSymbols.Keys);
-                aliases.Sort(StringComparer.Ordinal); // Actual order doesn't matter - just want to be deterministic.
+                aliases.Sort(StringComparers.IdentifierComparer); // Actual order doesn't matter - just want to be deterministic.
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/ImportChain.cs]: This is incorrect.

File: src/Compilers/Core/Portable/StringComparers.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+using System.Diagnostics;
+using System.Globalization;
+using System.Text;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    public static class StringComparers
+    {
```
Comment (gafter) [inline src/Compilers/Core/Portable/StringComparers.cs]: I don't see why we would need this type.

**Comment by gafter** [comment]
After finding all of the changes in the first 5-6 files to be incorrect, I stopped reviewing.

I don't know what the purpose of this PR is.

File: src/Workspaces/Core/Portable/Utilities/ObjectPools/SharedPools.cs [hunk 6]
```diff
@@ -45,7 +45,7 @@ public static ObjectPool<T> Default<T>() where T : class, new()
         /// pool that uses string as element with StringComparer.OrdinalIgnoreCase as element comparer
         /// </summary>
         public static readonly ObjectPool<HashSet<string>> StringIgnoreCaseHashSet =
-            new ObjectPool<HashSet<string>>(() => new HashSet<string>(StringComparer.OrdinalIgnoreCase), 20);
+            new ObjectPool<HashSet<string>>(() => new HashSet<string>(StringComparers.IdentifierComparer), 20);
```
Comment (gafter) [inline src/Workspaces/Core/Portable/Utilities/ObjectPools/SharedPools.cs]: >IdentifierComparer [](start = 86, length = 18)

This seems to violate the doc comment.

**Comment by CyrusNajmabadi** [comment]
Most of the feature changes are not appropraite.  

1. In the case where they should be checking identifiers, they should be using ISyntaxFactsService.StringComparer.
2. We should likely rename ISyntaxFactsService.StringComparer to ISyntaxFactsService.IdentifierComparer.
3. We should validate that ISyntaxFactsService.StringComaprer/IdentifierComparer returns the right value under the covers.

@gafter .  The IDE currently tries to use CaseInsensitiveComparison.Comparer to compare VB identifiers, and StringComparer.Ordinal for C# identifiers.  Is that correct?  Or should we be using different comparers over there?

**Comment by gafter** [comment]
That is correct. For file name comparisons use OrdinalIgnoreCase.

---
In reply to: [380222829](https://github.com/dotnet/roslyn/pull/26056#issuecomment-380222829) [](ancestors = 380222829)

**Comment by CyrusNajmabadi** [comment]
Great.  Thanks.  @AdamSpeight2008 if you want to fix up the IDE side, can you please do it as an independent PR.  And please use the approach i specified in my comment.   Once you do that, feel free to ping me and i can take a look. Thanks!

**Pull request closed without merging**